lib/observability/tracing/tracing.go

```diff
@@ -0,0 +1,211 @@
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, cfg.DialTimeout)
+	defer cancel()
+	exporter, err := otlptrace.New(ctx, traceClient)
+	switch {
+	case errors.Is(err, context.DeadlineExceeded):
+		return nil, trace.ConnectionProblem(err, "failed to connect to tracing exporter %q: %v", cfg.ExporterURL, err)
+	case err != nil:
+		return nil, trace.NewAggregate(err, traceClient.Stop(context.Background()))
+	}
+
+	return exporter, nil
+}
+
+// Provider wraps the OpenTelemetry tracing provider to provide common tags for all tracers.
+type Provider struct {
+	provider *sdktrace.TracerProvider
+}
+
+// Tracer returns a Tracer with the given name and options. If a Tracer for
+// the given name and options does not exist it is created, otherwise the
+// existing Tracer is returned.
+//
+// If name is empty, DefaultTracerName is used instead.
+//
+// This method is safe to be called concurrently.
+func (p *Provider) Tracer(instrumentationName string, opts ...oteltrace.TracerOption) oteltrace.Tracer {
+	opts = append(opts, oteltrace.WithInstrumentationVersion(teleport.Version))
+
+	return p.provider.Tracer(instrumentationName, opts...)
+}
+
+// Shutdown shuts down the span processors in the order they were registered.
+func (p *Provider) Shutdown(ctx context.Context) error {
+	return trace.Wrap(p.provider.ForceFlush(ctx), p.provider.Shutdown(ctx))
+}
+
+// NewTraceProvider creates a new Provider that is configured per the provided Config.
+func NewTraceProvider(ctx context.Context, cfg Config) (*Provider, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	exporter, err := NewExporter(ctx, cfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	attrs := []attribute.KeyValue{
+		// the service name used to display traces in backends
+		semconv.ServiceNameKey.String(cfg.Service),
+		attribute.String("teleport.version", teleport.Version),
+	}
+	attrs = append(attrs, cfg.Attributes...)
+
+	res, err := resource.New(ctx,
+		resource.WithFromEnv(),
+		resource.WithProcess(),
+		resource.WithTelemetrySDK(),
+		resource.WithHost(),
+		resource.WithAttributes(attrs...),
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// set global propagator, the default is no-op.
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{}))
+
+	// override the global logging handled with one that uses the
+	// configured logger instead
+	otel.SetErrorHandler(otel.ErrorHandlerFunc(func(err error) {
+		cfg.Logger.Warnf("failed to export traces: %v", err)
+	}))
+
+	// set global provider to our provider wrapper to have all tracers use the common TracerOptions
+	provider := &Provider{provider: sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(sdktrace.ParentBased(sdktrace.TraceIDRatioBased(cfg.SampleRatio))),
+		sdktrace.WithResource(res),
+		sdktrace.WithSpanProcessor(sdktrace.NewBatchSpanProcessor(exporter)),
+	)}
+	otel.SetTracerProvider(provider)
```

Comment:
Could we perhaps run the tracing configuration earlier than other components, and store this as a variable in `TeleportProcess` rather than storing it globally? Or is otel really only supposed to work with a single global exporter for the whole program?